scripts/lib/CIME/case_st_archive.py

```diff
@@ -62,6 +89,22 @@ def _get_ninst_info(case, compclass):
     logger.debug("ninst and ninst_strings are: {} and {} for {}".format(ninst, ninst_strings, compclass))
     return ninst, ninst_strings
 
+###############################################################################
+def _get_component_archive_entries(case, archive):
+###############################################################################
+    """
+    Each time this is generator function is called, it yields a tuple
+    (archive_entry, compname, compclass) for one component in this
+    case's compset components.
+    """
+    compset_comps = case.get_compset_components()
+    compset_comps.append('cpl')
+    compset_comps.append('dart')
+
+    for archive_entry in archive.get_entries():
```

Comment:
I think that this code is backward, instead of looping over all of the entries in env_archive and looking for a match in compname, one should just call self.get_optional_node('comp_archive_spec', attributes={"compname":compname}) for each compname.